### DISCUSSION TITLE
How can I execute raw draw commands?

### DISCUSSION THREAD

**Discussion by SanicBTW** [ppy/osu-framework#6308] [Q&A] [answered]
Hello everyone!
I am currently working on a port of the rhythm game known as Friday Night Funkin made in HaxeFlixel to osu!framework and I'm having some issues.

As most of you already know, sustains can have patterns, and my current implementation of sustains are just adding them to a buffered container and scale it to match the chart length but that will break the pattern from the graphic.

Alternatively I tried to look into osu!lazer source code to get an idea on how I could properly do sustains but I didn't really understand anything at all so I have resorted to my last option and that would be executing raw draw commands on the renderer.

Currently the sustains that are made in FNF are rendered using drawing triangles using some vertices and uv data from the sustain graphic, I'm wondering if I could at least do the same in this framework by using the renderer and draw methods that are available.

Here is the texture that is used to draw the sustains in-game:
![sustain](https://github.com/ppy/osu-framework/assets/58010536/448b2727-fd1e-47d7-b091-3cbaf88c1dda)

Any help is welcome!

**Comment by smoogipoo** [comment, marked as answer]
The general idea is to create a `DrawNode`. The song and dance looks something like this:

```csharp
public partial class MyDrawable : Drawable
{
    protected override DrawNode CreateDrawNode() => new MyDrawNode(this);
}

public class MyDrawNode : DrawNode
{
    public MyDrawNode(IDrawable source)
        : base(source)
    {
    }

    public override void ApplyState()
    {
        base.ApplyState();

        // Copy any states from the drawable locally here.
        // Typically, you'll need to cast Source to your drawable type somewhere to achieve this.
    }

    protected override void Draw(IRenderer renderer)
    {
        base.Draw(renderer);

        // Do your custom drawing here.
    }
}
```

From here, everything you need should be available in some form via the `IRenderer` param. There's other examples around the place demonstrating how to create [vertex buffers](https://github.com/ppy/osu-framework/blob/ec2c5595231ee9f81caf0f0ea03fb496f4e0bfd6/osu.Framework/Graphics/Audio/WaveformGraph.cs#L319-L378) or [uniform buffers](https://github.com/ppy/osu-framework/blob/ec2c5595231ee9f81caf0f0ea03fb496f4e0bfd6/osu.Framework/Graphics/UserInterface/CircularBlob.cs#L88-L170). You can do a project-wide search for inheritors of `DrawNode` for other examples.

Hope that helps :)

**Reply by SanicBTW** [reply]
That makes the thing, thank you so much!
I'll play around a lil bit around the functions I have to use to correctly render the stuff I want
